### ISSUE TITLE
`LIMIT` does not get pushed into subquery as expected

### ISSUE THREAD

**Issue by justin2004** [ontop/ontop#824]
when i send this query
```sparql
SELECT * WHERE {
?s a cco:Person .
} limit 1
```
to the `ontop/reformulate` endpoint

i see this SQL is generated
```sql
SELECT "v5"."src0m2" AS "src0m2"
FROM ((SELECT "v1"."src" AS "src0m2"
FROM "relationships" "v1"
WHERE ("v1"."src" IS NOT NULL AND 'COL_HERE' = "v1"."type")
LIMIT 1)UNION ALL 
(SELECT "v3"."dst" AS "src0m2"
FROM "relationships" "v3"
WHERE ("v3"."dst" IS NOT NULL AND 'COL_HERE' = "v3"."type")
LIMIT 1)) "v5"
LIMIT 1
```

notice that the `limit 1` gets pushed down into each sub query, as expected.

now if i do:
```sparql
SELECT * WHERE {
?s a cco:Person .
} limit 2
```

but notice that in the reformulation the `limit 2` is not pushed down to the subqueries as expected:
```sql
SELECT DISTINCT "v5"."src0m2" AS "src0m2"
FROM ((SELECT "v1"."src" AS "src0m2"
FROM "relationships" "v1"
WHERE ("v1"."src" IS NOT NULL AND 'COL_HERE' = "v1"."type")
)UNION ALL
(SELECT "v3"."dst" AS "src0m2"
FROM "relationships" "v3"
WHERE ("v3"."dst" IS NOT NULL AND 'COL_HERE' = "v3"."type")
)) "v5"
LIMIT 2
```

i have a table with billions of rows and when that limit does not get pushed down the query takes over 10 mins or does not finish.

shouldn't the limit (slice) get pushed into the sql subqueries?

**Comment by bcogrel** [comment]
Hi @justin2004, thanks for the suggestion.

The key observation here is that there is also a `distinct`. The correct way to push down `limit 2` would be to push down the `distinct` as well. Otherwise, without a `distinct`, I think the `limit 2` would have been pushed down.

I think it makes sense to extent the limit push-down to this interaction with `distinct`.

One special property of `limit 1` is that it eliminates distincts. I think it is what makes your query fast, as distincts are very costly when dealing with billions of rows. 

Would be it possible for you to specify unique constraints over your table to having distincts added to the mapping?
If you cannot modify the table, please consider adding a lens and specifying your mapping over it: https://ontop-vkg.org/tutorial/lenses/basic-lens.html#adding-constraints .

**Comment by justin2004** [comment]
> The key observation here is that there is also a distinct.

I don't use the `distinct` keyword in my sparql queries above. What `distinct` are you referring to?

**Comment by bcogrel** [comment]
Yes, I mean the `distinct` in the reformulated SQL. 

It comes from the virtual RDF graph. `distinct`s  are introduced by the mapping processor to make sure that the virtual graph is effectively a set of triples, i.e. it does not contain duplicates.

**Comment by justin2004** [comment]
@bcogrel are you suggesting i should change mappings like this
```
mappingId   MAPPING-ID3
target      ex:entity/{id_source} ex:is_supervised_by ex:entity/{id_destination} ; a ex:Person  .
source      SELECT * FROM "relationships" where
            "type" = 'SUPERVISOR'
```

to mappings like this?
```
mappingId   MAPPING-ID3
target      ex:entity/{id_source} ex:is_supervised_by ex:entity/{id_destination} ; a ex:Person  .
source      SELECT distinct id_source,id_destination FROM "relationships" where
            "type" = 'SUPERVISOR'
```

will that allow the distinct to get pushed into the SQL subqueries (when i use one in a sparql select)?

**Comment by bcogrel** [comment]
No, please do not insert a `distinct` in the mapping entry. Ontop will insert it automatically if needed when processing the mapping. Also its SQL parser doesn't yet support them, forcing it to treat the full SQL query as a "black-box view", preventing it from optimising it.

The goal is to avoid `distinct`s if possible, by declaring that `id_source` is unique (or the pair `id_source` and `id_destination` is unique) if that is the case.

For that, let's create a [basic lens](https://ontop-vkg.org/guide/advanced/lenses.html#basiclens) (assuming `id_source` is unique)

```json
{
    "relations": [
        {
            "name": ["\"lenses\"","\"relationships\""],
            "baseRelation": ["\"relationships\""],
            "uniqueConstraints": {
                "added": [
                    {
                        "name": "uc1",
                        "determinants": ["\"id_source\""]
                    }
                ]
            }
            "type": "BasicLens"
        }
    ]
}
```

and let's have the mapping entry pointing to it

```
mappingId   MAPPING-ID3
target      ex:entity/{id_source} ex:is_supervised_by ex:entity/{id_destination} ; a ex:Person  .
source      SELECT * FROM "lenses"."relationships" where
            "type" = 'SUPERVISOR'
```
